pkg/collector/corechecks/servicediscovery/module/envs.go

```diff
@@ -61,55 +44,103 @@ func readEnvsFile(path string) ([]string, error) {
 	if err != nil {
-		return nil
+		log.Warnf("error extracting injected envs: %s", err)
+		return nil, false
 	}
+	return injectionMeta, true
 
-	for _, entry := range entries {
-		path := filepath.Join(fdsPath, entry.Name())
-		name, err := os.Readlink(path)
-		if err != nil {
-			continue
-		}
+}
 
-		if name != injectorMemFdPath {
-			continue
-		}
+func extractInjectionMeta(path string) (*InjectedProcess, error) {
+	reader, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer reader.Close()
 
-		envs, _ := readEnvsFile(path)
-		return envs
+	data, err := io.ReadAll(io.LimitReader(reader, memFdMaxSize))
+	if err != nil {
+		return nil, err
+	}
+	if len(data) == memFdMaxSize {
+		return nil, io.ErrShortBuffer
 	}
 
-	return nil
+	var injectedProc InjectedProcess
+	_, err = injectedProc.UnmarshalMsg(data)
+	if err != nil {
+		return nil, err
+	}
+	return &injectedProc, nil
 }
 
-// envsToMap splits a list of strings containing environment variables of the
-// format NAME=VAL to a map.
-func envsToMap(envs ...string) map[string]string {
-	envMap := make(map[string]string, len(envs))
-	for _, env := range envs {
-		name, val, found := strings.Cut(env, "=")
-		if !found {
+func findInjectorFile(proc *process.Process) (string, bool) {
+	fdsPath := kernel.HostProc(strconv.Itoa(int(proc.Pid)), "fd")
+	// quick path, the shadow file is the first opened file by the process
+	// unless there are inherited fds
+	path := filepath.Join(fdsPath, "3")
+	if isInjectorFile(path) {
+		return path, true
+	}
+	entries, err := os.ReadDir(fdsPath)
+	if err != nil {
+		return "", false
+	}
+	for _, entry := range entries {
+		switch entry.Name() {
+		case "0", "1", "2", "3":
 			continue
+		default:
+			path := filepath.Join(fdsPath, entry.Name())
+			if isInjectorFile(path) {
+				return path, true
+			}
 		}
+	}
+	return "", false
+}
 
-		envMap[name] = val
+func isInjectorFile(path string) bool {
+	name, err := os.Readlink(path)
+	if err != nil {
+		return false
 	}
+	return strings.HasPrefix(name, injectorMemFdPath)
+}
 
-	return envMap
+// addEnvToMap splits a list of strings containing environment variables of the
+// format NAME=VAL to a map.
+func addEnvToMap(env string, envs map[string]string) {
+	name, val, found := strings.Cut(env, "=")
+	if found {
+		envs[name] = val
+	}
 }
 
 // getEnvs gets the environment variables for the process, both the initial
 // ones, and if present, the ones injected via the auto injector.
 func getEnvs(proc *process.Process) (map[string]string, error) {
-	envs, err := proc.Environ()
+	procEnvs, err := proc.Environ()
 	if err != nil {
 		return nil, err
 	}
-
-	envs = append(envs, getInjectedEnvs(proc)...)
-	return envsToMap(envs...), nil
+	envs := make(map[string]string, len(procEnvs))
+	for _, env := range procEnvs {
+		addEnvToMap(env, envs)
+	}
+	injectionMeta, ok := getInjectionMeta(proc)
+	if !ok {
+		return envs, nil
+	}
+	for _, env := range injectionMeta.InjectedEnv {
+		addEnvToMap(string(env), envs)
```

Comment:
If we're going to end up cutting at the `=` anyway we could just serialize {key, value} separately in the injector

Reply:
agreed let's pair and change it tomorrow